sdlib/d/gc/slab.d

```diff
@@ -23,6 +25,159 @@ struct SlabAllocGeometry {
 	}
 }
 
+struct SlabAllocInfo {
+public:
+	bool hasMetaData = false;
+
+	this(PageDescriptor pd, const void* ptr) {
+		assert(pd.isSlab(), "Expected a slab!");
+
+		e = pd.extent;
+		sg = SlabAllocGeometry(pd, ptr);
+		allowsMetaData = sizeClassSupportsMetadata(pd.sizeClass);
+		hasMetaData = allowsMetaData && e.hasMetaData(sg.index);
+	}
+
+	@property
+	auto address() {
+		return sg.address;
+	}
+
+	@property
+	size_t slotCapacity() {
+		return sg.size;
+	}
+
+	@property
+	size_t usedCapacity() {
+		return allowsMetaData ? sg.size - freeSpace : 0;
+	}
+
+	bool setUsedCapacity(size_t size) {
+		if (!allowsMetaData || (size > slotCapacity)) {
+			return false;
+		}
+
+		setFreeSpace(sg.size - size);
+		return true;
+	}
+
+private:
+	SlabAllocGeometry sg;
+	Extent* e;
+	bool allowsMetaData = false;
+
+	void setMetaData(bool value) {
+		assert(allowsMetaData, "size class not supports slab metadata!");
+
+		if (hasMetaData == value) {
+			return;
+		}
+
+		hasMetaData = value;
+		if (hasMetaData) {
+			e.enableMetaData(sg.index);
+			return;
+		}
+
+		e.disableMetaData(sg.index);
```

Comment:
It'd be one of these cases where if/else is warranted. Shouldn't the parameter be compiler time since it's always called with a constant?